sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/ThroughputEstimator.java

```diff
@@ -55,25 +46,18 @@ public ThroughputEstimator() {
    * @param timeOfRecords the committed timestamp of the records
    * @param bytes the total bytes of the records
    */
+  @SuppressWarnings("nullness") // queue is never null, nor the peeked element
   public void update(Timestamp timeOfRecords, long bytes) {
     synchronized (queue) {
-      BigDecimal bytesNum = BigDecimal.valueOf(bytes);
-      if (startTimeOfCurrentWindow.equals(Timestamp.MIN_VALUE)) {
-        bytesInCurrentWindow = bytesNum;
-        startTimeOfCurrentWindow = timeOfRecords;
-        return;
-      }
-
-      if (timeOfRecords.getSeconds() < startTimeOfCurrentWindow.getSeconds() + 1) {
-        bytesInCurrentWindow = bytesInCurrentWindow.add(bytesNum);
+      if (queue.isEmpty() || timeOfRecords.getSeconds() > queue.peek().getLeft().getSeconds()) {
+        queue.add(new ImmutablePair<>(timeOfRecords, BigDecimal.valueOf(bytes)));
       } else {
-        queue.add(new ImmutablePair<>(startTimeOfCurrentWindow, bytesInCurrentWindow));
-        bytesInQueue = bytesInQueue.add(bytesInCurrentWindow);
-
-        bytesInCurrentWindow = bytesNum;
-        startTimeOfCurrentWindow = timeOfRecords;
+        final ImmutablePair<Timestamp, BigDecimal> pair = queue.remove();
+        final ImmutablePair<Timestamp, BigDecimal> updatedPair =
+            new ImmutablePair<>(pair.getLeft(), pair.getRight().add(BigDecimal.valueOf(bytes)));
```

Comment:
Would this cause a memory issue if we keep creating new pairs when there are a lot of incoming records?

Reply:
Hum it is hard to say, the old pair would be garbage collected, but that is a good point. Let me see if I can use a mutable pair instead.

Reply:
I'm also a bit confused by the logic here.

Are you trying to comparing the incoming record with the most recently added or the earliest element? And in which case do you create a new element vs. append to the most recently added/earliest element?

Reply:
Changed to use a mutable class.

We are comparing the incoming record with the most recent timestamp. We create a new element if a second has passed from the most recent timestamp, otherwise we add to the bytes of the current element.

Reply:
One thing is: one ReadChangeStreamPartition SDF always corresponds to one change stream partition right? Wouldn't timestamps always be increasing for a change stream partition? So the incoming record should technically always be greater than the most recent timestamp?

Reply:
Sorry I should have clarified that we bucket the throughtput by seconds. So every record reported within the same timestamp **second** as the most recent element will result in just adding the bytes to that element. If a second has passed, then we will add a new element in the queue.